src/backends/torch/torchlib.cc

```diff
@@ -648,17 +648,67 @@ namespace dd
                                 cutout_params, geometry_params);
       }
 
+    Tensor class_weights = {};
+
+    if (ad_mllib.has("class_weights"))
+      {
+        std::vector<double> cwv
+            = ad_mllib.get("class_weights").get<std::vector<double>>();
+        if (cwv.size() != _nclasses)
+          {
+            this->_logger->error("class weights given, but number of "
+                                 "weights {} do not match "
+                                 "number of classes {}, ignoring",
+                                 cwv.size(), _nclasses);
+          }
+        else
+          {
+            this->_logger->info("using class weights");
+            auto options = torch::TensorOptions().dtype(torch::kFloat64);
+            class_weights
+                = torch::from_blob(cwv.data(), { _nclasses }, options)
+                      .to(torch::kFloat32)
+                      .to(_main_device);
+          }
+      }
+
     // solver params
     int64_t iterations = 1;
     int64_t batch_size = 1;
     int64_t iter_size = 1;
     int64_t test_batch_size = 1;
     int64_t test_interval = 1;
     int64_t save_period = 0;
-    TorchLoss tloss(_loss, _module.has_model_loss(), _seq_training, _timeserie,
-                    _regression, _classification, _module, this->_logger);
-    TorchSolver tsolver(_module, tloss, _devices, this->_logger);
-    Tensor class_weights = {};
+
+    typedef struct
+    {
+      std::shared_ptr<TorchModule> module;
+      std::shared_ptr<TorchLoss> loss;
+      std::shared_ptr<TorchSolver> solver;
+    } Rank;
```

Comment:
curious: why is this called Rank ? :)

Reply:
With ProcessGroups a "rank" is given to each process so that they know on which GPU they must operate. I kept the same name.